File: AndroidApp/shemsu-shop/app/src/main/java/com/example/data/local/LocalDb.kt
package com.example.data.local

import android.content.Context
import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Entity(tableName = "cart_items")
data class CartItemEntity(
    @PrimaryKey val name: String,
    val price: Double,
    val qty: Int,
    val isCustom: Boolean = false,
    val comment: String = ""
)

@Dao
interface CartDao {
    @Query("SELECT * FROM cart_items")
    fun getCartItems(): Flow<List<CartItemEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertItem(item: CartItemEntity)

    @Update
    suspend fun updateItem(item: CartItemEntity)

    @Query("DELETE FROM cart_items WHERE name = :name")
    suspend fun deleteItem(name: String)

    @Query("DELETE FROM cart_items")
    suspend fun clearCart()
}

@Database(entities = [CartItemEntity::class], version = 1, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {
    abstract fun cartDao(): CartDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "shemsu_shop_db"
                ).build()
                INSTANCE = instance
                instance
            }
        }
    }
}

File: AndroidApp/shemsu-shop/flutterApp/lib/data/local_db.dart
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import '../models/models.dart';

class LocalDb {
  static final LocalDb _instance = LocalDb._internal();
  static Database? _database;

  factory LocalDb() => _instance;

  LocalDb._internal();

  Future<Database> get database async {
    if (_database != null) return _database!;
    _database = await _initDb();
    return _database!;
  }

  Future<Database> _initDb() async {
    final dbPath = await getDatabasesPath();
    final path = join(dbPath, 'shemsu_shop_db.db');

    return await openDatabase(
      path,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE cart_items (
            name TEXT PRIMARY KEY,
            price REAL NOT NULL,
            qty INTEGER NOT NULL,
            isCustom INTEGER NOT NULL,
            comment TEXT NOT NULL
          )
        ''');
      },
    );
  }

  Future<List<CartItem>> getCartItems() async {
    final db = await database;
    final List<Map<String, dynamic>> maps = await db.query('cart_items');

    return List.generate(maps.length, (i) {
      return CartItem.fromJson(maps[i]);
    });
  }

  Future<void> insertItem(CartItem item) async {
    final db = await database;
    await db.insert(
      'cart_items',
      item.toJson(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  Future<void> updateItem(CartItem item) async {
    final db = await database;
    await db.update(
      'cart_items',
      item.toJson(),
      where: 'name = ?',
      whereArgs: [item.name],
    );
  }

  Future<void> deleteItem(String name) async {
    final db = await database;
    await db.delete(
      'cart_items',
      where: 'name = ?',
      whereArgs: [name],
    );
  }

  Future<void> clearCart() async {
    final db = await database;
    await db.delete('cart_items');
  }
}
